cannot use mockCNI (type *MockCNI) as type libcni.CNI

With github.com/containernetworking/cni 0.7.0 and Golang 1.12.2:

```
Testing    in: /builddir/build/BUILD/go-cni-891c2a41e18144b2d7921f971d6c9789a68046b2/_build/src
         PATH: /builddir/build/BUILD/go-cni-891c2a41e18144b2d7921f971d6c9789a68046b2/_build/bin:/builddir/.local/bin:/builddir/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin
       GOPATH: /builddir/build/BUILD/go-cni-891c2a41e18144b2d7921f971d6c9789a68046b2/_build:/usr/share/gocode
  GO111MODULE: off
      command: go test -buildmode pie -compiler gc -ldflags "-X github.com/containerd/go-cni/version.commit=891c2a41e18144b2d7921f971d6c9789a68046b2 -X github.com/containerd/go-cni/version=0 -extldflags '-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld '"
      testing: github.com/containerd/go-cni
github.com/containerd/go-cni
# github.com/containerd/go-cni [github.com/containerd/go-cni.test]
./namespace.go:31:32: not enough arguments in call to n.cni.AddNetworkList
	have (*libcni.NetworkConfigList, *libcni.RuntimeConf)
	want (context.Context, *libcni.NetworkConfigList, *libcni.RuntimeConf)
./namespace.go:39:29: not enough arguments in call to n.cni.DelNetworkList
	have (*libcni.NetworkConfigList, *libcni.RuntimeConf)
	want (context.Context, *libcni.NetworkConfigList, *libcni.RuntimeConf)
./cni_test.go:50:20: cannot use mockCNI (type *MockCNI) as type libcni.CNI in assignment:
	*MockCNI does not implement libcni.CNI (wrong type for AddNetwork method)
		have AddNetwork(*libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
		want AddNetwork(context.Context, *libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
./cni_test.go:66:20: cannot use mockCNI (type *MockCNI) as type libcni.CNI in assignment:
	*MockCNI does not implement libcni.CNI (wrong type for AddNetwork method)
		have AddNetwork(*libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
		want AddNetwork(context.Context, *libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
./cni_test.go:122:20: cannot use mockCNI (type *MockCNI) as type libcni.CNI in assignment:
	*MockCNI does not implement libcni.CNI (wrong type for AddNetwork method)
		have AddNetwork(*libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
		want AddNetwork(context.Context, *libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
./cni_test.go:123:20: cannot use mockCNI (type *MockCNI) as type libcni.CNI in assignment:
	*MockCNI does not implement libcni.CNI (wrong type for AddNetwork method)
		have AddNetwork(*libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
		want AddNetwork(context.Context, *libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
./cni_test.go:151:20: cannot use mockCNI (type *MockCNI) as type libcni.CNI in assignment:
	*MockCNI does not implement libcni.CNI (wrong type for AddNetwork method)
		have AddNetwork(*libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
		want AddNetwork(context.Context, *libcni.NetworkConfig, *libcni.RuntimeConf) (types.Result, error)
FAIL	github.com/containerd/go-cni [build failed]
```

--- a/cni.go
+++ b/cni.go
@@ -17,6 +17,7 @@
 package cni
 
 import (
+	"context"
 	"fmt"
 	"strings"
 	"sync"
@@ -29,9 +30,9 @@ import (
 
 type CNI interface {
 	// Setup setup the network for the namespace
-	Setup(id string, path string, opts ...NamespaceOpts) (*CNIResult, error)
+	Setup(ctx context.Context, id string, path string, opts ...NamespaceOpts) (*CNIResult, error)
 	// Remove tears down the network of the namespace.
-	Remove(id string, path string, opts ...NamespaceOpts) error
+	Remove(ctx context.Context, id string, path string, opts ...NamespaceOpts) error
 	// Load loads the cni network config
 	Load(opts ...CNIOpt) error
 	// Status checks the status of the cni initialization
@@ -139,7 +140,7 @@ func (c *libcni) Networks() []*Network {
 }
 
 // Setup setups the network in the namespace
-func (c *libcni) Setup(id string, path string, opts ...NamespaceOpts) (*CNIResult, error) {
+func (c *libcni) Setup(ctx context.Context, id string, path string, opts ...NamespaceOpts) (*CNIResult, error) {
 	if err := c.Status(); err != nil {
 		return nil, err
 	}
@@ -149,7 +150,7 @@ func (c *libcni) Setup(id string, path string, opts ...NamespaceOpts) (*CNIResul
 	}
 	var results []*current.Result
 	for _, network := range c.Networks() {
-		r, err := network.Attach(ns)
+		r, err := network.Attach(ctx, ns)
 		if err != nil {
 			return nil, err
 		}
@@ -159,7 +160,7 @@ func (c *libcni) Setup(id string, path string, opts ...NamespaceOpts) (*CNIResul
 }
 
 // Remove removes the network config from the namespace
-func (c *libcni) Remove(id string, path string, opts ...NamespaceOpts) error {
+func (c *libcni) Remove(ctx context.Context, id string, path string, opts ...NamespaceOpts) error {
 	if err := c.Status(); err != nil {
 		return err
 	}
@@ -168,7 +169,7 @@ func (c *libcni) Remove(id string, path string, opts ...NamespaceOpts) error {
 		return err
 	}
 	for _, network := range c.Networks() {
-		if err := network.Remove(ns); err != nil {
+		if err := network.Remove(ctx, ns); err != nil {
 			// Based on CNI spec v0.7.0, empty network namespace is allowed to
 			// do best effort cleanup. However, it is not handled consistently
 			// right now:

--- a/cni_test.go
+++ b/cni_test.go
@@ -17,12 +17,13 @@
 package cni
 
 import (
+	"context"
 	"net"
 	"testing"
 
 	cnilibrary "github.com/containernetworking/cni/libcni"
 	"github.com/containernetworking/cni/pkg/types"
-	"github.com/containernetworking/cni/pkg/types/020"
+	types020 "github.com/containernetworking/cni/pkg/types/020"
 	"github.com/containernetworking/cni/pkg/types/current"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
@@ -79,13 +80,15 @@ func TestLibCNIType020(t *testing.T) {
 	}, nil)
 	mockCNI.On("DelNetworkList", l.networks[1].config, expectedRT).Return(nil)
 
-	r, err := l.Setup("container-id1", "/proc/12345/ns/net")
+	ctx := context.Background()
+
+	r, err := l.Setup(ctx, "container-id1", "/proc/12345/ns/net")
 	assert.NoError(t, err)
 	assert.Contains(t, r.Interfaces, "eth0")
 	assert.NotNil(t, r.Interfaces["eth0"].IPConfigs)
 	assert.Equal(t, r.Interfaces["eth0"].IPConfigs[0].IP.String(), "10.0.0.1")
 	assert.Equal(t, r.Interfaces["eth0"].IPConfigs[1].IP.String(), "10.0.0.2")
-	err = l.Remove("container-id1", "/proc/12345/ns/net")
+	err = l.Remove(ctx, "container-id1", "/proc/12345/ns/net")
 	assert.NoError(t, err)
 
 	c := l.GetConfig()
@@ -174,7 +177,8 @@ func TestLibCNITypeCurrent(t *testing.T) {
 	}, nil)
 	mockCNI.On("DelNetworkList", l.networks[1].config, expectedRT).Return(nil)
 
-	r, err := l.Setup("container-id1", "/proc/12345/ns/net")
+	ctx := context.Background()
+	r, err := l.Setup(ctx, "container-id1", "/proc/12345/ns/net")
 	assert.NoError(t, err)
 	assert.Contains(t, r.Interfaces, "eth0")
 	assert.NotNil(t, r.Interfaces["eth0"].IPConfigs)
@@ -182,7 +186,7 @@ func TestLibCNITypeCurrent(t *testing.T) {
 	assert.Contains(t, r.Interfaces, "eth1")
 	assert.NotNil(t, r.Interfaces["eth1"].IPConfigs)
 	assert.Equal(t, r.Interfaces["eth1"].IPConfigs[0].IP.String(), "10.0.0.2")
-	err = l.Remove("container-id1", "/proc/12345/ns/net")
+	err = l.Remove(ctx, "container-id1", "/proc/12345/ns/net")
 	assert.NoError(t, err)
 }
 
@@ -190,22 +194,53 @@ type MockCNI struct {
 	mock.Mock
 }
 
-func (m *MockCNI) AddNetwork(net *cnilibrary.NetworkConfig, rt *cnilibrary.RuntimeConf) (types.Result, error) {
+func (m *MockCNI) AddNetwork(ctx context.Context, net *cnilibrary.NetworkConfig, rt *cnilibrary.RuntimeConf) (types.Result, error) {
 	args := m.Called(net, rt)
 	return args.Get(0).(types.Result), args.Error(1)
 }
 
-func (m *MockCNI) DelNetwork(net *cnilibrary.NetworkConfig, rt *cnilibrary.RuntimeConf) error {
+func (m *MockCNI) DelNetwork(ctx context.Context, net *cnilibrary.NetworkConfig, rt *cnilibrary.RuntimeConf) error {
 	args := m.Called(net, rt)
 	return args.Error(0)
 }
 
-func (m *MockCNI) DelNetworkList(net *cnilibrary.NetworkConfigList, rt *cnilibrary.RuntimeConf) error {
+func (m *MockCNI) DelNetworkList(ctx context.Context, net *cnilibrary.NetworkConfigList, rt *cnilibrary.RuntimeConf) error {
 	args := m.Called(net, rt)
 	return args.Error(0)
 }
 
-func (m *MockCNI) AddNetworkList(net *cnilibrary.NetworkConfigList, rt *cnilibrary.RuntimeConf) (types.Result, error) {
+func (m *MockCNI) AddNetworkList(ctx context.Context, net *cnilibrary.NetworkConfigList, rt *cnilibrary.RuntimeConf) (types.Result, error) {
 	args := m.Called(net, rt)
 	return args.Get(0).(types.Result), args.Error(1)
+}
+
+func (m *MockCNI) CheckNetworkList(ctx context.Context, net *cnilibrary.NetworkConfigList, rt *cnilibrary.RuntimeConf) error {
+	args := m.Called(net, rt)
+	return args.Error(0)
+}
+
+func (m *MockCNI) CheckNetwork(ctx context.Context, net *cnilibrary.NetworkConfig, rt *cnilibrary.RuntimeConf) error {
+	args := m.Called(net, rt)
+	return args.Error(0)
+}
+
+func (m *MockCNI) GetNetworkCachedResult(net *cnilibrary.NetworkConfig, rt *cnilibrary.RuntimeConf) (types.Result, error) {
+	args := m.Called(net, rt)
+	return args.Get(0).(types.Result), args.Error(1)
+}
+
+func (m *MockCNI) ValidateNetworkList(ctx context.Context, net *cnilibrary.NetworkConfigList) ([]string, error) {
+	args := m.Called(net)
+	return args.Get(0).([]string), args.Error(1)
+}
+
+func (m *MockCNI) ValidateNetwork(ctx context.Context, net *cnilibrary.NetworkConfig) ([]string, error) {
+	args := m.Called(net)
+	return args.Get(0).([]string), args.Error(1)
+}
+
+func (m *MockCNI) GetNetworkListCachedResult(net *cnilibrary.NetworkConfigList, rt *cnilibrary.RuntimeConf) (types.Result, error) {
+	args := m.Called(net, rt)
+	return args.Get(0).(types.Result), args.Error(1)
+
 }

--- a/namespace.go
+++ b/namespace.go
@@ -17,6 +17,8 @@
 package cni
 
 import (
+	"context"
+
 	cnilibrary "github.com/containernetworking/cni/libcni"
 	"github.com/containernetworking/cni/pkg/types/current"
 )
@@ -27,16 +29,16 @@ type Network struct {
 	ifName string
 }
 
-func (n *Network) Attach(ns *Namespace) (*current.Result, error) {
-	r, err := n.cni.AddNetworkList(n.config, ns.config(n.ifName))
+func (n *Network) Attach(ctx context.Context, ns *Namespace) (*current.Result, error) {
+	r, err := n.cni.AddNetworkList(ctx, n.config, ns.config(n.ifName))
 	if err != nil {
 		return nil, err
 	}
 	return current.NewResultFromResult(r)
 }
 
-func (n *Network) Remove(ns *Namespace) error {
-	return n.cni.DelNetworkList(n.config, ns.config(n.ifName))
+func (n *Network) Remove(ctx context.Context, ns *Namespace) error {
+	return n.cni.DelNetworkList(ctx, n.config, ns.config(n.ifName))
 }
 
 type Namespace struct {

--- a/vendor.conf
+++ b/vendor.conf
@@ -2,5 +2,5 @@ github.com/stretchr/testify b89eecf5ca5db6d3ba60b237ffe3df7bafb7662f
 github.com/davecgh/go-spew 8991bc29aa16c548c550c7ff78260e27b9ab7c73
 github.com/pmezard/go-difflib 792786c7400a136282c1664665ae0a8db921c6c2
 github.com/stretchr/objx 8a3f7159479fbc75b30357fbc48f380b7320f08e
-github.com/containernetworking/cni 142cde0c766cd6055cc7fdfdcb44579c0c9c35bf
+github.com/containernetworking/cni v0.7.1
 github.com/pkg/errors v0.8.0

